### PR TITLE
Outline for training and loss function

### PR DESCRIPTION
**Pull request by sophieml**
This is my attempt at an outline for the Viterbi loss and training loop (I haven't implemented the testing/decoding loop yet). I've done some minimal testing, but I have a couple standing questions before I refine things further:

- I wrote the `get_window_q` function to deal with OOB indexing when considering the search window, which zero-pads the output to the full search window size. This works for generating tau, but I also have to deal with OOB checking within the actual loss function, where we don't want to pad the output. I'm thinking the best way to do this is to generate the coordinates of the bounding box for the relevant values, and then slice that input/insert accordingly in the loss.
- I changed the utility functions so that they generate results for all q in Q besides `tau`. Because of this, the transition match probabilities for the entire image are computed separately in `tau_loop`. I'm not sure if it's worth bringing this function in line with others, given that it would involve changes to the search window function and the matches are only computed once per batch.
- I'm still thinking about a way to exclude the transcription padding from the loss. This is tricky because the loss is modified in two places (in the h/w loop, as well when adding the normalized trigram probability). My first thought is to sum the trigram tensor crosswise to create a (b, 1) mask and use that somehow. Since we also have to avoid updating the indices, it might be best to operate on the masked subset of rlp?
- (edited to add) I also need to add conversion logic from the boundary coordinates with respect to the receptive field, to absolute coordinates in order to work with the current implementation of `epsilon`. Or is this necessary? Instead of subtracting the absolute q with the meshgrid, can we do the same calculation and subtract the center of the receptive field as a constant?

### CODE REVIEW

**Comment by nhprof** [comment]
Responses:
1.  Can we make get_window_q return the coordinates as well as the window values?  We could use those further down in the loss function.  I think that is what you also suggested.
2.  I'm fine with leaving tau_loss a separate function.  There's no practical difference, is there?
3.  For the transcription padding, I have two thoughts.  We may not need both of them.  Thought #1 is that we could redefine tlp for slices that have reached the end of their transcript.  The new tlp would be inf everywhere except for 0 at (+1,0).  This would mean that the backtrace variables would behave in a predictable way, and the loss would not change.  But it doesn't deal with the normalized component that is added later.  Thought #2 might work there.  We can make a mask tensor that will be 1 everywhere we have a real trigram, and 0 where the trigram is (0,0,0).  If we multiply the added piece by this mask, then again the loss won't change.  I think that's probably what we want.  However, it doesn't really save us any work; all the computation still happens.  It makes me wonder whether using larger batch sizes is really worth it, because there is so much computation potentially wasted.
4. I'm not sure if I understand this comment.  I'll think about it some more.  Also see comments below.

Comments/Questions:
1.  I get an error in the computation of C_L_diff and L_C_diff.
2. The dimensions of the cropped regions should be 2wh+1 by ww, since we want to be symmetric around the center pixel.
3. I think there's a missing outdent at the bottom of viterbi_loss?
4.  I think the epsilon computation needs indices in the assignment; it currently assigns the x part to the whole.  Also bdot needs two arguments as written.  And maybe it wants a squeeze() at the end?

**Comment by sophieml** [comment]
I'm in agreement with points 1 and 2. Some thoughts about 3:

I'm not exactly sure what you mean by (+1, 0). Do you mean the value at that tensor index is zero? I think we'd have to create a Boolean mask to do this regardless—I can't think of a way to selectively index the input otherwise.  If we subset the batch dimension, I think we can dodge updating the indices?
`rlp[mask, g+1, i-wh:i+wh+1, j+1:j+ww] = current[mask]` (etc)
I'd have to add null checks/skips to the decoding part, since there'd be empty entries, but I think this could work in combination with thought #2.

On 4: Since the model itself only takes in the basic unit, the sample boundaries will be relative to that unit. This causes a problem, since B_C at q_1 and B_L at q_2 won't align (B_L is to the left of B_C in relative coordinates, but has the same absolute coordinates). I need to incorporate the original coordinates from the outer loops into the sample boundaries somehow—am I underthinking things, or is simply adding +(i, j) enough?

I'll take a look at your other comments tomorrow, let me know if you have further questions.

**Comment by nhprof** [comment]
Yes, I think you're right.  Let's make a mask and use it in both updates -- then my comment about (+1,0) is obsolete, and it also saves us computation.  I'm not sure whether we'll also need an antimask to update the values for the completed trigrams -- I think we can leave those empty, and just ignore them when we interpret the results at the end.  Am I right that as currently written, viterbi_loss computes and returns a loss for every starting pixel -- it doesn't seem to include the call to min2d at the end of my matlab code, correct?

Cancel my comment #1 above; I was calling with bad arguments.

I see what you're worrying about with the relative vs. absolute for the boundary coordinates.  In my thinking I had been assuming that B used absolute coordinates, but if it's coming straight out of the network computation then it is probably relative to the basic unit.  I think the program is written as though we were both thinking about absolute coordinates, so if B contains all relative numbers, then one way to deal with it is to convert B to Babs right at the start, making whatever change is necessary.  I kind of like doing it that way because it compartmentalizes the process so that we can think about each step separately.

**Comment by sophieml** [comment]
I think we do need an anti-mask—since we initialize tlp[g+1] to be inf, we won't be able to backtrack if we don't bring the values from the last trigram forward (after the h/w iteration). I think I have a typo in my min calculation—it should be `loss, idx = rlp[:, n_trigrams+2].view(-1, h*w).min(dim=-1)` —the next two lines retrieve the (i, j) coordinates from the linear `idx`.

**Comment by nhprof** [comment]
I see -- and it also wants to be outdented so it is outside the g loop.

**Comment by sophieml** [comment]
Merging this, I'll create a new branch for the alternative formulation if necessary.
